### PR TITLE
file: Support unicode urls on windows

### PR DESCRIPTION
**Pull request by foopoiuyt**
Support file: urls with utf-8 filenames on Windows when _UNICODE is defined.

### CODE REVIEW

**Comment by jay** [comment]
https://dev.azure.com/daniel0244/5571c33c-81e1-43d7-93ee-d3d4152f27b2/_apis/build/builds/4396/logs/99

~~~
2021-01-21T00:23:08.4699191Z file.c: In function 'file_upload':
2021-01-21T00:23:08.4700646Z file.c:280:55: error: macro "open" passed 3 arguments, but takes just 2
2021-01-21T00:23:08.4701344Z    fd = open(file->path, mode, data->set.new_file_perms);
2021-01-21T00:23:08.4701886Z                                                        ^
2021-01-21T00:23:08.4704068Z file.c:280:6: error: assignment to 'int' from 'int (__attribute__((cdecl)) *)(const char *, int,  ...)' {aka 'int (*)(const char *, int,  ...)'} makes integer from pointer without a cast [-Wint-conversion]
2021-01-21T00:23:08.4705100Z    fd = open(file->path, mode, data->set.new_file_perms);
2021-01-21T00:23:08.4705770Z       ^
~~~

That is because [open](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/open-wopen?view=msvc-160) takes 2 or 3 arguments. To fix it you could do `#  define open          curlx_win32_open` and then make curlx_win32_open variadic. The alternative is a variadic macro but I don't think that's supported with all compilers.

File: lib/curl_setup.h [hunk 1]
```diff
@@ -356,9 +358,11 @@
 #    define fstat(fdes,stp)            _fstat(fdes, stp)
 #    define stat(fname,stp)            curlx_win32_stat(fname, stp)
 #    define struct_stat                struct _stat
+#    define open(fname,oflag)          curlx_win32_open(fname, oflag)
```
Comment (emilengler) [inline lib/curl_setup.h:361]: Please don't redfine a POSIX syscall

File: lib/curl_multibyte.c [hunk 2]
```diff
@@ -82,6 +82,21 @@ char *curlx_convert_wchar_to_UTF8(const wchar_t *str_w)
...
+  int result = 0;
+  wchar_t *filename_w = curlx_convert_UTF8_to_wchar(filename);
+  if(filename_w)
+    result = _wopen(filename_w, oflag);
+  free(filename_w);
+  if(result)
+    return result;
+#endif
+
+  return (open)(filename, oflag);
```
Comment (emilengler) [inline lib/curl_multibyte.c:97]: Why even define the function on non Windows platforms?

**Review by emilengler — CHANGES_REQUESTED** [review]

File: lib/curl_multibyte.c [hunk 3]
```diff
@@ -82,6 +82,21 @@ char *curlx_convert_wchar_to_UTF8(const wchar_t *str_w)
 
 #if defined(USE_WIN32_LARGE_FILES) || defined(USE_WIN32_SMALL_FILES)
 
+int curlx_win32_open(const char *filename, int oflag)
+{
+#ifdef _UNICODE
+  int result = 0;
```
Comment (MarcelRaad) [inline lib/curl_multibyte.c:88]: The error value is defined as -1 and 0 might be a valid result: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/open-wopen?view=msvc-160#return-value

File: lib/curl_multibyte.c (hunk 2 above)
Comment (MarcelRaad) [reply lib/curl_multibyte.c:97]: It's not, this is in a `#if defined(USE_WIN32_LARGE_FILES) || defined(USE_WIN32_SMALL_FILES)` block.

File: lib/curl_multibyte.c [hunk 4]
```diff
@@ -82,6 +82,29 @@ char *curlx_convert_wchar_to_UTF8(const wchar_t *str_w)
 
 #if defined(USE_WIN32_LARGE_FILES) || defined(USE_WIN32_SMALL_FILES)
 
+int curlx_win32_open(const char *filename, int oflag, ...)
+{
+    int pmode = 0;
```
Comment (MarcelRaad) [inline lib/curl_multibyte.c:87]: Indentation is wrong here.

File: lib/curl_multibyte.c [hunk 5]
```diff
@@ -82,6 +82,29 @@ char *curlx_convert_wchar_to_UTF8(const wchar_t *str_w)
...
+        pmode = va_arg(param, int);
+    va_end(param);
+
+#ifdef _UNICODE
+  int result = -1;
+  wchar_t *filename_w = curlx_convert_UTF8_to_wchar(filename);
+  if(filename_w)
+    result = _wopen(filename_w, oflag, pmode);
+  free(filename_w);
+  if(result)
```
Comment (MarcelRaad) [inline lib/curl_multibyte.c:101]: This should be `if(result != -1)` now.

File: lib/curl_multibyte.c (hunk 3 above)
Comment (foopoiuyt) [reply lib/curl_multibyte.c]: Thanks, totally missed it (and the later check as mentioned below).

File: lib/curl_setup.h [hunk 6]
```diff
@@ -356,9 +359,12 @@
 #    define fstat(fdes,stp)            _fstat(fdes, stp)
 #    define stat(fname,stp)            curlx_win32_stat(fname, stp)
 #    define struct_stat                struct _stat
+#    define open(fname,oflag,...)      curlx_win32_open(fname, oflag \
+                                       , ## __VA_ARGS__)
```
Comment (jay) [inline lib/curl_setup.h:363]: I would do away with __VA_ARGS__ I don't think all compilers support it

**Review by jay — APPROVED** [review]

**Review by nardl — APPROVED** [review]

**Review by MarcelRaad — APPROVED** [review]

**Comment by jay** [comment]
Thanks

**Pull request closed without merging**